File: lib/src/main/java/io/github/kory33/guardedqueries/core/fol/FunctionFreeSignature.java
package io.github.kory33.guardedqueries.core.fol;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import io.github.kory33.guardedqueries.core.utils.extensions.FormulaExtensions;
import uk.ac.ox.cs.gsat.GTGD;
import uk.ac.ox.cs.pdq.fol.ConjunctiveQuery;
import uk.ac.ox.cs.pdq.fol.Formula;
import uk.ac.ox.cs.pdq.fol.Predicate;

import java.util.Collection;
import java.util.stream.Collectors;

/**
 * An object of this class represents a first-order logic signature with
 * - countably-infinitely many constants
 * - finite set of predicate symbols
 * - no function symbols
 */
public record FunctionFreeSignature(ImmutableSet<Predicate> predicates) {
    public FunctionFreeSignature(Iterable<? extends Predicate> predicates) {
        this(ImmutableSet.copyOf(predicates));
    }

    public static FunctionFreeSignature fromFormulas(final Collection<? extends Formula> formulas) {
        return new FunctionFreeSignature(
                formulas
                        .stream()
                        .flatMap(FormulaExtensions::streamPredicatesAppearingIn)
                        .collect(Collectors.toList())
        );
    }

    public static FunctionFreeSignature encompassingRuleQuery(
            final Collection<? extends GTGD> rules,
            final ConjunctiveQuery query
    ) {
        return FunctionFreeSignature.fromFormulas(
                ImmutableList
                        .<Formula>builder()
                        .addAll(rules)
                        .add(query)
                        .build()
        );
    }

    public ImmutableSet<String> predicateNames() {
        return ImmutableSet.copyOf(predicates.stream().map(Predicate::getName).toList());
    }

    public int maxArity() {
        return predicates.stream()
                .mapToInt(Predicate::getArity)
                .max()
                .orElse(0);
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/fol/FunctionFreeSignature.scala
package io.github.kory33.guardedqueries.core.fol

import io.github.kory33.guardedqueries.core.utils.extensions.FormulaExtensions
import uk.ac.ox.cs.gsat.GTGD
import uk.ac.ox.cs.pdq.fol.{ConjunctiveQuery, Formula, Predicate}
import io.github.kory33.guardedqueries.core.utils.extensions.FormulaExtensions.given

/**
 * An object of this class represents a first-order logic signature with
 *   - countably-infinitely many constants
 *   - finite set of predicate symbols
 *   - no function symbols
 */
object FunctionFreeSignature {
  private def fromFormulas(formulas: Set[Formula]) =
    new FunctionFreeSignature(formulas.flatMap(_.allPredicates))

  def encompassingRuleQuery(rules: Set[GTGD], query: ConjunctiveQuery): FunctionFreeSignature =
    FunctionFreeSignature.fromFormulas(Set(query: Formula) ++ rules)
}

case class FunctionFreeSignature(predicates: Set[Predicate]) {
  def this(predicates: Iterable[Predicate]) = {
    this(predicates.toSet)
  }

  def predicateNames: Set[String] =
    predicates.map(_.getName)

  def maxArity: Int = predicates.map(_.getArity) match {
    case arities if arities.isEmpty => 0
    case arities                    => arities.max
  }
}
